107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java

```diff
@@ -128,34 +172,38 @@ Eh107CacheManager getCacheManager(ConfigSupplier configSupplier, Properties prop
           byURI.remove(uri, cacheManager);
         }
 
-        cacheManager = createCacheManager(uri, configSupplier.getConfiguration(), properties);
+        cacheManager = cacheManagerCreator.get();
         byURI.put(uri, cacheManager);
       }
     }
 
     return cacheManager;
   }
 
-  private Eh107CacheManager createCacheManager(URI uri, Configuration config, Properties properties) {
-    Eh107CacheLoaderWriterProvider cacheLoaderWriterFactory = new Eh107CacheLoaderWriterProvider();
+  private Eh107CacheManager createCacheManager(URI uri, ClassLoader classLoader, CacheManagerBuilder<org.ehcache.CacheManager> builder) {
+    Set<Service> services = builder.getServices();
```

Comment:
This method feels like it contains logic that should really be a part of the CacheManagerBuilder.  Fixing this would mean either adding a new method to the builder that only added the service if it didn't already exist, or modifying the contract of the using method to behave more consistently with the using method that take a config object. This would really be moving a failure that will occur anyway when using `using` from being a CM build time to earlier in the builder invocations (arguably a good thing).

Reply:
What would you do instead? `usingIfNotPresent` or `using(Service service, boolean overrideIfPresent)`? I agree that I don't like the `getServices` either.

Reply:
I'm thinking the sensible solution would be to add `usingIfAbsent(...)` methods to the builder. At the same time it might make sense to also align the contracts for the service and service-config related methods since I think they vary a bit at the moment.